File: src/commands/group/create.ts
import {ErrorCodes, SDKError} from "@ironcorelabs/ironnode";
import {Command, flags as flagtype} from "@oclif/command";
import chalk from "chalk";
import * as GroupMaps from "../../lib/GroupMaps";
import {ironnode} from "../../lib/SDK";
import {keyFile} from "../../lib/sharedFlags";
import {buildCommandSampleText, createDisplayTable} from "../../lib/Utils";

/**
 * Group create command. Creates a new group with the user provided ID. Auto adds the creator of the group as a member and an admin.
 */
export default class Create extends Command {
    static description = `Create a new cryptographic group. Upon creation of the group you will become both an admin and a member of the group. This generates a public key for the group and uploads it to the IronCore service.`;
    static args = [
        {
            name: "group",
            description: "Name for the group. Will be used when referencing this group from all other commands.",
            required: true,
        },
    ];
    static flags = {
        help: flagtype.help({char: "h"}),
        keyfile: keyFile(),
    };
    static examples = [buildCommandSampleText("group:create myGroup", "Create a new group with the name 'myGroup'")];

    /**
     * Check if the user is already a member/admin of a group with the provided name and fail if they are or
     * if we can't get the list of groups they're in.
     */
    async checkGroupExists(groupName: string) {
        let doesGroupExist: boolean = false;
        try {
            doesGroupExist = await GroupMaps.doesGroupNameAlreadyExist(groupName);
        } catch (e) {
            this.log(e);
            this.error(chalk.red("Unable to make group create request."));
        }

        if (doesGroupExist) {
            this.error(chalk.red(`You are already in a group with the name '${groupName}'.`));
        }
    }

    async run() {
        const {args} = this.parse(Create);
        const groupName = args.group as string;
        if (groupName.includes(",") || groupName.includes("^")) {
            return this.error("Group names cannot contain commas or carets.");
        }
        await this.checkGroupExists(args.group);

        return ironnode()
            .group.create({groupName: args.group})
            .then((group) => {
                const table = createDisplayTable(["Group Name", "Group ID", "Admin", "Member"]);
                table.push([group.groupName, group.groupID, chalk.green("✔"), chalk.green("✔")]);

                this.log(chalk.green("\nNew group successfully created."));
                this.log(chalk.green(`\n${table.toString()}\n`));
            })
            .catch((e: SDKError) => {
                if (e.code === ErrorCodes.GROUP_CREATE_REQUEST_FAILURE) {
                    this.error(chalk.red("Unable to make group create request."));
                }
                this.error(chalk.red(e.message));
            });
    }
}

File: src/group/create.rs
use crate::util::{GetKeyfile, group_already_known, println_paint};
use clap::Parser;
use ironoxide::group::GroupName;
use ironoxide::prelude::BlockingIronOxide;
use yansi::Paint;

use ironoxide::IronOxideErr;
use std::convert::TryFrom;
use std::path::PathBuf;

const EXAMPLE: &str = "EXAMPLE

    Create a new group with the name 'myGroup'
        $ ironhide group create myGroup

";

#[derive(Parser)]
#[clap(after_help = EXAMPLE)]
/// Create a new cryptographic group. Upon creation of the group you will become both an admin and a member of the
/// group. This generates a public key for the group and uploads it to the IronCore service.
pub struct Create {
    /// Name for the group. Will be used when referencing this group from all other
    /// commands.
    #[clap(value_parser = parse_group_name)]
    name: GroupName,
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(value_parser = clap::value_parser!(PathBuf), short, long)]
    keyfile: Option<PathBuf>,
}

pub(crate) fn parse_group_name(s: &str) -> Result<GroupName, IronOxideErr> {
    GroupName::try_from(s)
}

impl GetKeyfile for Create {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

pub fn group_create(sdk: &BlockingIronOxide, Create { name, .. }: Create) -> Result<(), String> {
    if group_already_known(sdk, &name) {
        println_paint(Paint::red(format!(
            "You're already in a group with the name {}",
            name.name()
        )));
    } else if name.name().contains('^') || name.name().contains(',') {
        println_paint(Paint::red(
            "Group names cannot contain commas or carets.".to_string(),
        ));
    } else {
        // change this if https://github.com/IronCoreLabs/ironoxide/issues/263 ever is finished
        let options = ironoxide::group::GroupCreateOpts::new(
            None,
            Some(name.clone()),
            true,
            true,
            None,
            Vec::new(),
            Vec::new(),
            false,
        );
        match sdk.group_create(&options) {
            Ok(group) => {
                let mut table =
                    table!([Fbb->"Group Name", Fbb->"Group ID", Fbb->"Admin", Fbb -> "Member"]);
                table.add_row(
                    row![group.name().unwrap().name(), group.id().id(), Fg -> "✔", Fg -> "✔"],
                );
                println_paint(Paint::green("New group successfully created.".to_string()));
                table.printstd();
            }
            Err(err) => {
                println_paint(Paint::red(format!(
                    "Failed to create group {}: {}",
                    name.name(),
                    err
                )));
            }
        };
    }

    Ok(())
}
